File: _src/main/java/raccoonlang/parser/ast/DataClassBodyAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.TokenType;

import java.util.ArrayList;
import java.util.List;

public class DataClassBodyAstNode {
    public List<DataClassMemberAstNode> dataClassMemberAstNodes = new ArrayList<>();

    public static DataClassBodyAstNode parse(Parser parser) {
        DataClassBodyAstNode dataClassAstNode = new DataClassBodyAstNode();

        parser.take(TokenType.OPEN_CURLY);

        while(true){
            DataClassMemberAstNode dataClassMemberAstNode = DataClassMemberAstNode.tryParse(parser);
            if(dataClassMemberAstNode != null) {
                dataClassAstNode.dataClassMemberAstNodes.add(dataClassMemberAstNode);
            }else{
                break;
            }
        }

        parser.take(TokenType.CLOSE_CURLY);

        return dataClassAstNode;
    }
}

File: Parsing/AST/DataClassBodyAstNode.cs
namespace Raccoonlang.Parsing.AST;

using Tokenizing;

public class DataClassBodyAstNode
{
    public List<DataClassMemberAstNode> MemberList { get; set; } = new();

    public static DataClassBodyAstNode Parse(Parser parser)
    {
        DataClassBodyAstNode node = new DataClassBodyAstNode();

        if (parser.Peek().Type == TokenType.Semicolon)
        {
            parser.Skip();
            return node;
        }

        parser.Take(TokenType.OpenCurly);
        while (parser.Peek().Type != TokenType.CloseCurly)
        {
            node.MemberList.Add(DataClassMemberAstNode.Parse(parser));
        }

        return node;
    }

    public override string ToString() => $"DataClassBodyAstNode={{MemberList=[{string.Join(",", MemberList)}]}}";
}